tests/upgrade-tests/utils.py

```diff
@@ -116,3 +121,48 @@ def all_units_ready(model):
     while not all_units_ready(model):
         assert_no_unit_errors(model)
         await asyncio.sleep(1)
+
+async def conjureup(namespace, bundle, channel, snap_channel, model):
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        cmd = 'charm pull --channel=%s cs:~%s/%s %s'
+        cmd %= channel, namespace, bundle, os.path.join(tmpdirname, bundle)
+        cmd = cmd.split()
+        process = await asyncio.create_subprocess_exec(*cmd)
+        await process.wait()
+        assert process.returncode == 0
+        shutil.copytree(
+            os.path.join('/snap/conjure-up/current/spells', bundle),
+            os.path.join(tmpdirname, 'spell')
+        )
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-01_get-kubectl'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-01_get-kubectl.yaml'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-02_cluster-info'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-02_cluster-info.yaml'))
+        with open(os.path.join(tmpdirname, bundle, 'bundle.yaml')) as f:
+            bundledata = yaml.load(f)
+        appkey = 'services' if 'services' in bundledata else 'applications'
+        master = bundledata[appkey]['kubernetes-master']
+        worker = bundledata[appkey]['kubernetes-worker']
+        master.setdefault('options', {})['channel'] = snap_channel
+        worker.setdefault('options', {})['channel'] = snap_channel
```

Comment:
We'll need to make `snap_channel` optional. If we're just testing the latest charms, for example, we probably want to stick with the snap channels that are specified in the bundle.